Objects/object.c

```diff
@@ -526,9 +526,17 @@ PyObject_CallFinalizerFromDealloc(PyObject *self)
      * never happened. */
     _Py_ResurrectReference(self);
 
-    _PyObject_ASSERT(self,
-                     (!_PyType_IS_GC(Py_TYPE(self))
-                      || _PyObject_GC_IS_TRACKED(self)));
+    if (_PyObject_IS_GC(self)) {
+        /* If we get here, it must have originally been tracked when
+         * _Py_Dealloc was called.  Make it tracked again so it can't become
+         * part of a reference cycle and leak.  Some types expect the object
+         * to be tracked when this returns non-zero.
+         */
+        if (!_PyObject_GC_IS_TRACKED(self)) {
+            _PyObject_GC_TRACK(self);
+        }
```

Comment:
Is it safe to rely on this? The `Py_DECREF` used in the interpreter does not call `_Py_Dealloc`:

https://github.com/python/cpython/blob/35541c410d894d4fa18002f7fdbebfe522f8320e/Python/ceval.c#L87-L88

Reply:
Oh, good catch. I'd say it's not safe given that.  I believe `Py_DECREF` used to always call `_Py_Dealloc` but that's no longer the case.  Adding my trashcan logic to this code path is going to be a bit of a perf hit (mostly due to the IS_GC check).

Reply:
I think I've fixed this by having the Py_DECREF macro in ceval.c call `_Py_Dealloc`.  That might be too much of a perf hit though.  A bit cheaper would be to do the GC untrack if required but not check the trashcan depth from that macro.